### PR TITLE
Add dataset directory stream method to dataset manager

### PR DESCRIPTION
**Pull request by dmgaldi**
## Overview
* Adding `streamAllDatasets(): Stream<DatasetDirectory>` to DatasetManager`
* Added an eagerly loaded implementation of `DatasetDirectory` since we have access to all objects in S3
* In support of above, added alternate constructors to all `*FileImpl` classes that takes an object instead of s3 location info

## Testing
Just unit tests for now. There's some slightly tricky logic to group the dataset files together that could possibly be simplified?

### CODE REVIEW

File: src/main/kotlin/org/veupathdb/vdi/lib/s3/datasets/DatasetDeleteFlagFileImpl.kt [hunk 1]
```diff
@@ -1,8 +1,40 @@
...
+        loadObjectStream = { bucket.objects.open(path)?.stream }
+    )
+
+    constructor(s3Object: S3Object) : this(
+        path = s3Object.path,
+        lastModifiedSupplier = { s3Object.lastModified },
+        existsChecker = { true }, // It definitely exists if loaded from an actual S3 object
+        loadObjectStream = { s3Object.bucket.objects.open(s3Object.path)?.stream }
+    ) {
+        if (s3Object.baseName !== S3Paths.SHARE_OFFER_FILE_NAME) {
```
Comment (Foxcapades) [inline src/main/kotlin/org/veupathdb/vdi/lib/s3/datasets/DatasetDeleteFlagFileImpl.kt:33]: Why is the delete file looking for `SHARE_OFFER_FILE_NAME`, shouldn't it be the delete flag file?

File: src/main/kotlin/org/veupathdb/vdi/lib/s3/datasets/DatasetDirectoryImpl.kt [hunk 2]
```diff
@@ -8,6 +8,7 @@ import org.veupathdb.vdi.lib.common.model.*
 import org.veupathdb.vdi.lib.json.JSON
 import org.veupathdb.vdi.lib.s3.datasets.paths.S3DatasetPathFactory
 import java.io.InputStream
+import java.time.OffsetDateTime
```
Comment (Foxcapades) [inline src/main/kotlin/org/veupathdb/vdi/lib/s3/datasets/DatasetDirectoryImpl.kt:11]: Nitpick: Unnecessary import?

File: src/main/kotlin/org/veupathdb/vdi/lib/s3/datasets/DatasetManager.kt [hunk 3]
```diff
@@ -42,9 +48,125 @@ class DatasetManager(private val s3Bucket: S3Bucket) {
...
+            this.currentDataset = EagerlyLoadedDatasetDirectory(stagedObjects, idsFromObject.first, idsFromObject.second, pathFactory)
+            stagedObjects = emptyList()
+            return currentDataset
+          }
+          this.currentDataset = null
+          return null
+        }
+        val currDatasetIDs = datasetIdFromS3Object(stagedObjects.first())
+        val currUserID = currDatasetIDs.first
+        val currDatasetID = currDatasetIDs.second
```
Comment (Foxcapades) [inline src/main/kotlin/org/veupathdb/vdi/lib/s3/datasets/DatasetManager.kt:148]: Not suggesting a change, just pointing out a handy kotlin thing you can do here: unpacking the returned pair by doing:
```kotlin
val (currUserID, currDatasetID) = datasetIdFromS3Object(stagedObjects.first())
```

this also works for data classes and classes that implement the appropriate operator functions for unpacking

**Review by Foxcapades — CHANGES_REQUESTED** [review]
Looks good except a likely copy/paste issue in [DatasetDeleteFlagFileImpl.kt](https://github.com/VEuPathDB/vdi-component-s3/pull/1/files#diff-8c181d8c4776680a4e8ab36251648e758ab724746aba84a24a7d1de02e27f2fe)

File: src/main/kotlin/org/veupathdb/vdi/lib/s3/datasets/DatasetManager.kt (hunk 3 above)
Comment (dmgaldi) [reply src/main/kotlin/org/veupathdb/vdi/lib/s3/datasets/DatasetManager.kt:148]: I tried this and missed the parentheses! I thought it was just wishful thinking

File: src/main/kotlin/org/veupathdb/vdi/lib/s3/datasets/DatasetDeleteFlagFileImpl.kt (hunk 1 above)
Comment (dmgaldi) [reply src/main/kotlin/org/veupathdb/vdi/lib/s3/datasets/DatasetDeleteFlagFileImpl.kt:33]: Oops, I'll add a unit test with a delete flag.

**Review by Foxcapades — APPROVED** [review]

**Pull request merged**